src/controller/gc/controller.go

```diff
@@ -0,0 +1,208 @@
+package gc
+
+import (
+	"context"
+	"encoding/json"
+	"github.com/goharbor/harbor/src/jobservice/job"
+	"github.com/goharbor/harbor/src/lib/errors"
+	"github.com/goharbor/harbor/src/lib/log"
+	"github.com/goharbor/harbor/src/lib/q"
+	"github.com/goharbor/harbor/src/pkg/scheduler"
+	"github.com/goharbor/harbor/src/pkg/task"
+)
+
+var (
+	// GCCtl is a global garbage collection controller instance
+	GCCtl = NewController()
+)
+
+const (
+	// SchedulerCallback ...
+	SchedulerCallback = "GARBAGE_COLLECTION"
+	// gcVendorType ...
+	gcVendorType = "GARBAGE_COLLECTION"
+)
+
+// Controller manages the tags
+type Controller interface {
+	// Start start a manual gc job
+	Start(ctx context.Context, parameters map[string]interface{}) (int64, error)
+	// Stop stop a gc job
+	Stop(ctx context.Context, taskID int64) error
+	// GetLog get the gc log by id
+	GetLog(ctx context.Context, id int64) ([]byte, error)
+	// GetByID get gc executions per id
+	GetByID(ctx context.Context, id int64) (*History, error)
+	// History list all of gc executions
+	History(ctx context.Context, query *q.Query) ([]*History, error)
+	// Count count the gc executions
+	Count(ctx context.Context, query *q.Query) (int64, error)
+	// GetSchedule get the current gc schedule
+	GetSchedule(ctx context.Context) (*scheduler.Schedule, error)
+	// CreateSchedule create the gc schedule with cron string
+	CreateSchedule(ctx context.Context, cron string, parameters map[string]interface{}) (int64, error)
+	// DeleteSchedule remove the gc schedule
+	DeleteSchedule(ctx context.Context) error
+}
+
+// NewController creates an instance of the default repository controller
+func NewController() Controller {
+	return &controller{
+		taskMgr:      task.NewManager(),
+		exeMgr:       task.NewExecutionManager(),
+		schedulerMgr: scheduler.New(),
+	}
+}
+
+type controller struct {
+	taskMgr      task.Manager
+	exeMgr       task.ExecutionManager
+	schedulerMgr scheduler.Scheduler
+}
+
+// Start starts the manual GC
+func (c *controller) Start(ctx context.Context, parameters map[string]interface{}) (int64, error) {
+	execID, err := c.exeMgr.Create(ctx, gcVendorType, -1, task.ExecutionTriggerManual, parameters)
+	if err != nil {
+		return -1, err
+	}
+	taskID, err := c.taskMgr.Create(ctx, execID, &task.Job{
+		Name: job.ImageGC,
+		Metadata: &job.Metadata{
+			JobKind: job.KindGeneric,
+		},
+		Parameters: parameters,
+	})
+	if err != nil {
+		return -1, err
+	}
+	defer func() {
+		if err == nil {
+			return
+		}
+		if err := c.taskMgr.Stop(ctx, taskID); err != nil {
+			log.Errorf("failed to stop the task %d: %v", taskID, err)
+		}
+	}()
+	return execID, nil
+}
+
+// Stop ...
+func (c *controller) Stop(ctx context.Context, taskID int64) error {
```

Comment:
Use the executionID as the input of the function as the ID returned by `Start` is the execution ID